src/System.Net.Sockets/src/System/Net/Sockets/IOControlKeepAlive.cs

```diff
@@ -0,0 +1,103 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Net.Sockets
+{
+    internal class IOControlKeepAlive
+    {
+        private const uint WindowsDefaultTimeMs = 7200000u;
+        private const uint WindowsDefaultIntervalMs = 1000u;
+        private static readonly bool s_supportsKeepAliveViaSocketOption = SupportsKeepAliveViaSocketOption();
+        private static readonly ConditionalWeakTable<SafeSocketHandle, IOControlKeepAlive> s_socketKeepAliveCwt = new ConditionalWeakTable<SafeSocketHandle, IOControlKeepAlive>();
+        [ThreadStatic]
+        private static byte[] s_keepAliveValuesBuffer;
+
+        private uint _timeMs = WindowsDefaultTimeMs;
+        private uint _intervalMs = WindowsDefaultIntervalMs;
+
+        public static bool TryGet(SafeSocketHandle handle, SocketOptionLevel optionLevel, SocketOptionName optionName, out int optionValue)
+        {
+            if (IsNotNeeded(optionLevel, optionName))
+            {
+                optionValue = 0;
+                return false;
+            }
+                
+            IOControlKeepAlive ioControlKeepAlive;
+            s_socketKeepAliveCwt.TryGetValue(handle, out ioControlKeepAlive);
+            optionValue = optionName == SocketOptionName.TcpKeepAliveTime ?
+                MillisecondsToSeconds(ioControlKeepAlive?._timeMs ?? WindowsDefaultTimeMs) :
+                MillisecondsToSeconds(ioControlKeepAlive?._intervalMs ?? WindowsDefaultIntervalMs);
+            return true;
+        }
+
+        public static SocketError? Set(SafeSocketHandle handle, SocketOptionLevel optionLevel, SocketOptionName optionName, int optionValue)
```

Comment:
This wouldn't be to be nullable. I think it would be better if the SocketPal call site liked like:
if (option == ... && level == ... && IOControlKeepAlive.IsNeeded)
{
    return IOControlKeepAlive.Set(...);
}